### PR TITLE
Pic 2025

### PR DESCRIPTION
**Pull request by rschallapalli**

### CODE REVIEW

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/CourtCaseControllerIntTest.java [hunk 1]
```diff
@@ -656,4 +666,110 @@ void GET_cases_givenNoCreatedFilterParams_whenGetCases_thenReturnAllCases() {
         }
     }
 
+    @Nested
+    class OffenderEndpointTests {
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/controller/CourtCaseControllerIntTest.java:670]: You could consider breaking these out into a separate class. We started using Nested classes a while back but honestly I'm not sure it was a good idea, it just produces massive tests. Interested to hear what your thoughts are?

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/CourtCaseControllerIntTest.java [hunk 2]
```diff
@@ -656,4 +666,110 @@ void GET_cases_givenNoCreatedFilterParams_whenGetCases_thenReturnAllCases() {
...
+                .assertThat()
+                .statusCode(200);
+
+            final var offender = offenderRepository.findByCrn("Y320741").get();
+            assertThat(offender.getProbationStatus()).isEqualTo(OffenderProbationStatus.NOT_SENTENCED);
+            assertThat(offender.isBreach()).isFalse();
+        }
+
+        @Test
+        void PUT_offender_givenDefendantIdWithNoOffenderAndGivenNewOffender_whenPutOffender_theCreateAndAssocaiteOffender() {
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/controller/CourtCaseControllerIntTest.java:741]: I'm amazed I even spotted this typo 😆 

```suggestion
        void PUT_offender_givenDefendantIdWithNoOffenderAndGivenNewOffender_whenPutOffender_theCreateAndAssociateOffender() {
```

File: src/test/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+        var offender = offenderUpdateService.getDefendantOffenderByDefendantId(DEFENDANT_ID).block();
+        assertThat(offender).isEqualTo(DefendantOffender.builder().build());
+        verify(defendantRepository).findFirstByDefendantIdOrderByIdDesc(DEFENDANT_ID);
+        verify(offenderRepository).findByCrn(CRN);
+    }
+
+    @Test
+    void shouldReturnOffenderWhenFoundInDatabase() {
+        given(defendantRepository.findFirstByDefendantIdOrderByIdDesc(DEFENDANT_ID))
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateServiceTest.java:66]: Hmm 🤔  Quite like these BDDMockito forms 👍

**Review by jhevans — COMMENTED** [review]
A few comments and suggestions. As discussed we'll leave this one until I get back anyway!

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/model/DefendantOffender.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import java.time.LocalDate;
+import java.util.Optional;
+
+@Data
+@Builder
+@NoArgsConstructor(access = AccessLevel.PRIVATE, force = true)
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class DefendantOffender {
+    private final String probationStatus;
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/controller/model/DefendantOffender.java]: addressed

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateService.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    @Transactional
+    public void removeDefendantOffenderAssociation(final String defendantId) {
+        final var defendant = findDefendantOrElseThrow(defendantId);
+
+        if (!StringUtils.isEmpty(defendant.getCrn())) {
+            defendantRepository.save(defendant.withCrn(null).withId(null));
+        }
+    }
+
+    public Mono<DefendantOffender> getDefendantOffenderByDefendantId(final String defendantId) {
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateService.java]: addressed

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateService.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        final var defendant = findDefendantOrElseThrow(defendantId);
+
+        final var offender =  StringUtils.isEmpty(defendant.getCrn()) ? DefendantOffender.builder().build() :
+                offenderRepository.findByCrn(defendant.getCrn()).map(DefendantOffender::of).orElse(DefendantOffender.builder().build());
+
+        return Mono.just(offender);
+    }
+
+    @Transactional
+    public Mono<DefendantOffender> updateDefendantOffender(final String defendantId, OffenderEntity offenderUpdate) {
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateService.java]: addressed

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateService.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+        final var defendant = findDefendantOrElseThrow(defendantId);
+
+        final var dbOffender = offenderRepository.findByCrn(offenderUpdate.getCrn());
+
+        final var offenderToUpdate = dbOffender.map(oe -> {
+            oe.setProbationStatus(offenderUpdate.getProbationStatus());
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderUpdateService.java]: addressed

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/CourtCaseControllerIntTest.java (hunk 1 above)
Comment (rschallapalli) [reply src/test/java/uk/gov/justice/probation/courtcaseservice/controller/CourtCaseControllerIntTest.java]: moved as suggested

**Review by jhevans — APPROVED** [review]

**Review by jhevans — APPROVED** [review]

**Pull request merged**